dbuf-typecheck/src/strategy.rs

```
@@ -0,0 +1,226 @@
+        InternedTypeDeclaration,
+    },
+    simple::{SimpleTyper, SimpleTyperError},
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum CheckerTask {
+    Signature(InternedString),
+    Branch {
+        type_name: InternedString,
+        branch_index: usize,
+    },
+    Constuctor(InternedString),
+}
+
+#[derive(Debug)]
+pub struct StrategyBuilder {
+    typer: SimpleTyper,
+    signature_scope: HashMap<InternedString, InternedString>,
+    pattern_scope: HashMap<InternedString, InternedString>,
+    constructor_scope: HashMap<InternedString, InternedString>,
+    graph: TopSortBuilder<CheckerTask>,
+}
+
+#[derive(Error)]
+pub enum StrategyError<Str> {
+    #[error(transparent)]
+    SimpleTyperError(#[from] SimpleTyperError<Str>),
+    #[error("incorrect access operator")]
+    IncorrectAccess,
+    #[error("number of patterns isn't equal to number of dependencies")]
+    IncorrectNumberOfPatterns,
+}
+
+type InternedStrategyError = StrategyError<InternedString>;
+
+impl StrategyBuilder {
+    pub fn build_strategy<Loc>(
+        module: &InternedModule<Loc>,
+    ) -> Result<Vec<CheckerTask>, InternedStrategyError> {
+        let mut builder = Self {
+            typer: SimpleTyper::from_module(module),
+            signature_scope: Default::default(),
+            pattern_scope: Default::default(),
+            constructor_scope: Default::default(),
+            graph: TopSortBuilder::new(),
+        };
+        for definition in module {
+            builder.analyze_type(definition)?;
+        }
+        Ok(builder.graph.top_sort())
+    }
+
+    fn analyze_type<Loc>(
+        &mut self,
+        definition: &InternedTypeDeclaration<Loc>,
+    ) -> Result<(), InternedStrategyError> {
+        let dependencies: Vec<_> = definition
+            .dependencies
+            .iter()
+            .map(|dependency| {
+                let ExpressionNode::FunCall { fun: type_name, .. } = dependency.node else {
+                    todo!()
```

What would be done here?

---

I think I need to return an error here for now, but I didn't quite think it through so left it as TODO. I also can check it in SimpleTyper and just put `unreachable!()` here